[RISCV] Use separate CCValAssign for both parts of f64 with ilp32.

topperc opened the pull request:
Mark any registers as CustomReg.
    
This allows us to more directly emit the register or memory access for the high part. Previously we needed a memory access if the low register was X17 and we assumed the stack offset was 0. If the low part wasn't X17, we assumed the high register was the next register after the low register.
    
This is another part of supporting FP arguments with GISel.

topperc committed:
[RISCV] Use separate CCValAssign for both parts of f64 with ilp32.

Mark any registers as CustomReg.

This allows us to more directly emit the register or memory
access for the high part. Previously we assume the offset was
0 and we needed a memory access if the low register was X17.

This is another part of supporting FP arguments with GISel.

topperc committed:
!fixup Use getCustomMem for the high memory location.

asb commented:
This seems to introduce some compile failures in the GCC torture suite. e.g. pr44942 for rv32imafdc, ilp32, O1.

topperc committed:
!fixup Change use of i to OutIdx in LowerCall.

topperc commented:
> This seems to introduce some compile failures in the GCC torture suite. e.g. pr44942 for rv32imafdc, ilp32, O1.

Thanks, Alex! I've fixed that issue and rebased the patch on trunk to get rid of the already committed https://github.com/llvm/llvm-project/pull/69118.

asb reviewed (approved):
LGTM.
